core/src/test/java/org/elasticsearch/index/mapper/NumberFieldTypeTests.java

```diff
@@ -382,4 +386,51 @@ public void doTestDocValueRangeQueries(NumberType type, Supplier<Number> valueSu
         reader.close();
         dir.close();
     }
+
+    public void testParseOutOfRangeValues() throws IOException {
+        final List<OutOfRangeSpec<Object>> inputs = Arrays.asList(
+            OutOfRangeSpec.of(NumberType.HALF_FLOAT, "65504.1", "[half_float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.FLOAT, "3.4028235E39", "[float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.DOUBLE, "1.7976931348623157E309", "[double] supports only finite values"),
+
+            OutOfRangeSpec.of(NumberType.HALF_FLOAT, 65504.1, "[half_float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.FLOAT, 3.4028235E39, "[float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.DOUBLE, new BigDecimal("1.7976931348623157E309"), "[double] supports only finite values"),
+
+            OutOfRangeSpec.of(NumberType.HALF_FLOAT, Float.NaN, "[half_float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.FLOAT, Float.NaN, "[float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.DOUBLE, Double.NaN, "[double] supports only finite values"),
+
+            OutOfRangeSpec.of(NumberType.HALF_FLOAT, Float.POSITIVE_INFINITY, "[half_float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.FLOAT, Float.POSITIVE_INFINITY, "[float] supports only finite values"),
+            OutOfRangeSpec.of(NumberType.DOUBLE, Double.POSITIVE_INFINITY, "[double] supports only finite values")
+        );
+
+        for (OutOfRangeSpec<Object> item: inputs) {
+            try {
+                item.type.parse(item.value, false);
+                fail("Parsing exception expected for [" + item.type + "] with value [" + item.value + "]");
+            } catch (IllegalArgumentException e) {
+                assertThat("Incorrect error message for [" + item.type + "] with value [" + item.value + "]",
+                    e.getMessage(), containsString(item.message));
+            }
```

Comment:
Despite expectThrows is more concise and readable, I think we should keep try/fail/catch/assert because `fail("Mapper parsing exception expected for [" + item.type + "] with value [" + item.value + "]");` shows which **numbertype** with which **value** failed.